benchmark/single-source/FindStringNaive.swift

```diff
@@ -0,0 +1,116 @@
+//===--- FindStringNaive.swift --------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import TestsUtils
+
+// Mini benchmark implementing a naive String search algorithm that
+// at the moment shows a lot of ARC traffic.
+let t: [BenchmarkCategory] = [.String, .refcount]
+let N = 100
+
+var longStringXsThenFoo: String?
+var longArrayXsThenFoo: [UInt8]?
+
+public let FindStringNaive = [
+  BenchmarkInfo(
+    name: "FindStringNaive.Substring",
+    runFunction: runBenchSubstring,
+    tags: t,
+    setUpFunction: {
+      longStringXsThenFoo = String(repeating: "x", count: 10_000) + "foo <-- needle"
+    }),
+  BenchmarkInfo(
+    name: "FindStringNaive.String",
+    runFunction: runBenchString,
+    tags: t,
+    setUpFunction: {
+      longStringXsThenFoo = String(repeating: "x", count: 10_000) + "foo <-- needle"
+    }),
+  BenchmarkInfo(
+    name: "FindStringNaive.ArrayUTF8",
+    runFunction: runBenchArrayOfUTF8,
+    tags: t,
+    setUpFunction: {
+      longArrayXsThenFoo = Array(repeating: UInt8(ascii: "x"), count: 1_000_000) + Array("foo <-- needle".utf8)
+    }),
+]
+
+func find<S: StringProtocol>(_ string: S,
+                             needle1: Character,
+                             needle2: Character?,
+                             needle3: Character?) -> String.Index? {
+    var index = string.startIndex
+    while index < string.endIndex {
+        if string[index] == needle1 {
+            if let needle2 = needle2 {
+                if find(string[string.index(after: index)...],
+                        needle1: needle2,
+                        needle2: needle3,
+                        needle3: nil) == string.index(after: index) {
+                  return index
+                }
+            } else {
+                return index
+            }
+        }
+        index = string.index(after: index)
+    }
```

Comment:
This could be cleaner if rewritten to use `guard`, local variables, etc., without (hopefully) changing what the benchmark is attempting to measure (such as ARC surrounding the recursive call). E.g.

```swift
while index < string.endIndex {
  var nextIndex = string.index(after: index)
  if string[index] == needle1 {
    // Check subsequent needles recursively (if applicable)
    guard let needle2 = needle2 else { return index }

    if find(
      string[nextIndex...], needle1: needle2, needle2: needle3, needle3: nil
    ) == nextIndex {
      return index
    }
  }
  index = nextIndex
}
```